packages/runner/src/storage/interface.ts

```diff
@@ -174,6 +174,177 @@ export interface IStorageManagerV2 {
+ * value. Property name denotes type of notification.
+ */
+type StorageNotification =
+  | ICommitNotification
+  | IRevertNotification
+  | ILoadNotification
+  | IPullNotification
+  | IIntegrateNotification;
+
+/**
+ * This notification is broadcasted after commit on {@link IStorageTransaction}
+ * is called and underlying changes are written to the local replica. Note that
+ * this represents a local optimistic update which can be denied by remote
+ * storage provider in which case they will be reverted.
+ */
+interface ICommitNotification {
+  type: "commit";
+
+  /**
+   * The space into which changes were made.
+   */
+  space: MemorySpace;
+  /**
+   * Set of changes merged.
+   */
+  changes: IMergedChanges;
+  /**
+   * Transaction that committed changes.
+   */
+  source: IStorageTransaction;
+}
+
+/**
+ * This notification is broadcasted if commited changes were denied and had to
+ * be reverted.
+ */
+interface IRevertNotification {
+  type: "revert";
+
+  /**
+   * The space into which changes were made.
+   */
+  space: MemorySpace;
+  /**
+   * Set of changes merged. Note that this is not necessary resetting every
+   * change commit made to a state it had pre-commit as things may have changed
+   * since and `before` values will represent state in the replica before we
+   * reverted them to the state in the `after`. Also note that set of changes
+   * in the commit may be larger than set of changes here because commit may
+   * have being stacked on to of the other and if first commit was denied and
+   * reverted changes are some state may have already being updated by previous
+   * revert.
+   */
+  changes: IMergedChanges;
+  /**
+   * Transaction that committed changes.
+   */
+  source: IStorageTransaction;
+
+  /**
+   * Reason storage had to revert changes.
+   */
+  reason: StorageTransactionRejected;
+}
+
+/**
+ * This notification is broadcasted when storage loads changes from the local
+ * cache into a storage.
+ */
+interface ILoadNotification {
+  type: "load";
+  space: MemorySpace;
+  changes: IMergedChanges;
+}
+
+/**
+ * This notification is broadcasted when storage pulls changes from the remote
+ * storage provider and merges them into the local replica.
+ */
+interface IPullNotification {
+  type: "pull";
+  space: MemorySpace;
+  changes: IMergedChanges;
+}
+
+/**
+ * This notification is broadcasted after storage receives integrates changes from
+ * the remote storage provider into a local replica.
+ */
+interface IIntegrateNotification {
+  type: "integrate";
+  space: MemorySpace;
+  changes: IMergedChanges;
+}
+
+/**
+ * Set of changes that were merged into the local replica.
+ */
+interface IMergedChanges extends Iterable<IMemoryChange> {
+}
+
+interface IMemoryChange {
+  /**
+   * Memory address that was changed.
+   */
+  address: IMemoryAddress;
+  /**
+   * Value memory address had before change.
+   */
+  before?: JSONValue;
```

Comment:
this is ultra-nit, so mostly asking to confirm my understanding:

`undefined` implies retracted (or equivalently never set) in our system. so it's possible for something to go from JSONValue to `undefined` and that is correct. Is there a case where we instead want to say "we don't know"? I don't think we need the latter, since what we care about is that .get() returned something defined and then afterwards get() returns `undefined` (and same the other way around as well).

It would be cleaner to write this as `before: JSONValue | undefined` to convey this clearly, even though it means the same.